### PR TITLE
Add receipts to highscore

### PR DESCRIPTION
**Pull request by elmaxe**
- Made the high score list fancier (table)
- Added the ability to order by amount of receipts instead of just total amount.

Co-authored-by: Douglas Fischer <anakin_skryter@hotmail.se>

image
image

### CODE REVIEW

**Comment by oskstr** [comment]
When you show the number of receipts, wouldn’t that be “Flest Cashflowat”? 🤔

**Review by niklasvatn — COMMENTED** [review]
LGTM in general.
I don't know the system too well (or django and Vue).
The only thing I thought about is the text of the button. I think it makes more sense to display the text that you will change to.

@oskstr should make the final review.

File: templates/stats/index.html [hunk 1]
```diff
@@ -12,20 +12,67 @@ <h2>{{ year|intcomma }} kr</h2>
...
+                        {% for user in highscore_amount %}
+                            <tr>
+                                <td style="width: 5%;">{{ forloop.counter }}</td>
+                                <td style="width: 45%;"><a href="{% url 'user-show' user.user.username %}">{{ user.user.get_full_name }}</a></td>
+                                <td style="width: 25%;">{{ user.total_amount|intcomma }} kr</td>
+                                <td style="width: 25%;">{{ user.receipts }} st</td>
+                            </tr>
+                        {% endfor %}
+                </tbody>
+                <tbody v-if="selected == 'Antal cashflows'">
```
Comment (oskstr) [inline templates/stats/index.html:40]: Code could be a bit more DRY (don't repeat yourself) 😅 

I can imagine you thought about it already, but the styling is repeated twice here even though it is exactly the same. 

I'm not sure how you would conditionally loop over one of two lists using Jinja2 (`{% for x in y %}`) . But I'm pretty sure you could achieve it with Vue and a `v-for` loop. But if it is unnecessarily difficult, it is probably fine as is. You would probably be storing the profile lists like committees are stored in `templates/admin/expenses/overview.html`.

**Review by oskstr — APPROVED** [review]
LGTM 👍 

I'm posting a couple of comments that you could have a look at. But it seems to work as is imo.

File: stats/views.py [hunk 2]
```diff
@@ -15,8 +15,11 @@ def index(request):
     highscore = models.Profile.objects \
         .filter(expense__reimbursement__isnull=False, expense__expensepart__amount__lt=20000) \
         .annotate(total_amount=Sum('expense__expensepart__amount')) \
-        .filter(total_amount__gte=0) \
-        .order_by('-total_amount')[:10]
+        .annotate(receipts=Count('expense__expensepart')) \
+        .filter(total_amount__gte=0)
+
+    highscore_amount = highscore.order_by('-total_amount')[:10]
+    highscore_receipts = highscore.order_by('-receipts')[:10]
```
Comment (oskstr) [inline stats/views.py]: I'm not sure why but it seems like one of my review comments didn't show up. 🤷 Or I'm just not seeing it.

I suggested a secondary ordering, like:
```python
highscore_receipts = highscore.order_by('-receipts', '-total_amount')[:10]
```

That would resolve ties in receipts by highest total amount, I think. We are using it in some other places. Maybe have the same type of tiebreaker for amount.

File: templates/stats/index.html (hunk 1 above)
Comment (oskstr) [reply templates/stats/index.html:40]: Actually, I maybe just found a pretty good solution: [Macros](https://jinja.palletsprojects.com/en/3.0.x/templates/#macros). Seems to be pretty much exactly for this use case.

**Comment by oskstr** [comment]
@elmaxe I discovered an issue that this introduced btw. Basically only admins are allowed to view users so I get a 403 error page if I click on one. I know that there is an issue (#110) to actually create a 403 page which would make it a bit better. 

But even better UX in this case would probably be to only have link if user `may_be_viewed_by` the logged in user, which is true if it is yourself or you are an admin.

**Comment by elmaxe** [comment]
@oskstr the links to the users weren't introduced with this PR. They were already there from the beginning

**Comment by oskstr** [comment]
Ah, my bad 😅

**Pull request merged**